natlib/svcd.c

```diff
@@ -185,3 +192,201 @@ static int svcd_init( lua_State *L )
 
     return 0;
 }
+
+//////////////////////////////////////////////////////////////////////////////
+// SVCD.write((string targetip, number svcid, number attrid, lightfunction payload, 
+//                        number timeout_ms, lightfunction on_done)) 
+// Authors: Aparna Dhinakaran, David Ho, Romi Phadte, Cesar Torres
+/////////////////////////////////////////////////////////////
+
+
+void resolve_table(lua_State *L, char* key){
+    // push the global table, resolve it and then pop it off the stack
+    lua_getglobal(L, "SVCD");
+    lua_pushstring(L, key);
+    lua_gettable(L, -2);
+    lua_insert(L, -2);
+    lua_settop(L, -2);
+}
+
+static int svcd_write_invoke_later(lua_State *L);
+static int svcd_callback_handler(lua_State *L);
+static int anon_function(lua_State *L);
+
+static int svcd_callback_handler(lua_State *L){
+    int ivkd = (int) lua_tonumber(L, lua_upvalueindex(1));
+    printf("making callback function using ivkd = %i \n", ivkd);
+
+    resolve_table(L, "TIMEOUT");
+    int timeout = (int) lua_tonumber(L, -1);
+
+    resolve_table(L, "handlers"); // 7
```

Comment:
While it is a good abstraction for development, this may add extra overhead to the embedded system since SVCD need to be load/unload every time this function is called, but in reality, it only need to be load once. 

Untitled